src/core/qgsmaplayerregistry.cpp

```diff
@@ -151,7 +164,9 @@ void QgsMapLayerRegistry::removeMapLayers( const QList<QgsMapLayer*>& layers )
     QString myId( lyr->id() );
     emit layerWillBeRemoved( myId );
     emit layerWillBeRemoved( lyr );
+    mRegistryLock.lockForWrite();
     mMapLayers.remove( myId );
+    mRegistryLock.unlock();
```

Comment:
Emitting signal layerWillBeRemoved() from a thread may cause all sorts of trouble here:
- if a slot has queued connection - the layer will be deleted before the signal is delivered
- if a slot has direct connection - suddenly a code that was not meant to be called from thread will be run in worker thread

The same applies to layersAdded() signal I believe...